kvserver: deduplicate replica state used in replicate queue planning

AlexTalks opened the pull request:
Previously throughout the replicate queue planning stage represented by the recently introduced `replicateQueue.PlanOneChange(..)`, the range descriptor and span config was obtained multiple times from a replica, potentially introducing discrepancies and/or staleness as no lock is held on the replica state in the interim. This change modifies the replicate queue planning phase to use a single call to `repl.DescAndSpanConfig()`, the results of which are passed throughout as needed. The change also updates the naming and comments of some replicate queue functions used in the planning phase to indicate that they are intended to be without side effects.

Release note: None

AlexTalks committed:
kvserver: deduplicate replica state used in replicate queue planning

Previously throughout the replicate queue planning stage represented by
the recently introduced `replicateQueue.PlanOneChange(..)`, the range
descriptor and span config was obtained multiple times from a replica,
potentially introducing discrepancies and/or staleness as no lock is
held on the replica state in the interim. This change modifies the
replicate queue planning phase to use a single call to
`repl.DescAndSpanConfig()`, the results of which are passed throughout
as needed. The change also updates the naming and comments of some
replicate queue functions used in the planning phase to indicate that
they are intended to be without side effects.

Part of #91570.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/92172)
<!-- Reviewable:end -->

kvoli reviewed (commented):
Nice patch. This makes the code more logical and consistent.

There's one other spot in the allocator code it would be good to also drop in `ValidLeaseTargets`. Drop `Desc` from the adhoc interface in this fn in favor of providing a descriptor:

https://github.com/cockroachdb/cockroach/blob/1a6e9f885baa124d5ff2996adb966ea15a1a9b2b/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L1634-L1634

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92172#-:-NHqYpGN44ee5hLO4D4w:bbrn1f5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92172#-NHq_f0T0RGN5H4SrwTZ:-NHq_f0T0RGN5H4SrwT_:buq1ict) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7e54d176700707401226f43715f65ca3d60f48c/pkg/kv/kvserver/replicate_queue.go#L1220)):*
> ```Go
> func (rq *replicateQueue) planAddOrReplaceVoters(
> 	ctx context.Context,
> 	repl *Replica,
> ```

Checking the calls to this fn and others that take a repl arg, it looks like we are getting closer to being able to drop the repl argument completely.

That seems like a worthwhile goal, should we file an issue to do just this? Moving these functions into a separate pkg would help with clarity into the future.


<!-- Sent from Reviewable.io -->

AlexTalks committed:
kvserver: always pass range descriptor to allocator as argument

While the allocator works, for the most part, by using the range
descriptor as passed in by callers (i.e. Replicate Queue, Store
Rebalancer, Simulator), in a few cases where leaseholder replicas are
passed in the allocator was looking up the range descriptor from the
replica. This change makes the range descriptor parameter uniform, to
ensure that it is specified by the caller.

Part of #91570.

Release note: None

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92172#-NHq_f0T0RGN5H4SrwTZ:-NJD-8688nxj_5j4GcTw:b-n4j64l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7e54d176700707401226f43715f65ca3d60f48c/pkg/kv/kvserver/replicate_queue.go#L1220)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Checking the calls to this fn and others that take a repl arg, it looks like we are getting closer to being able to drop the repl argument completely.

That seems like a worthwhile goal, should we file an issue to do just this? Moving these functions into a separate pkg would help with clarity into the future.
</blockquote></details>

Could also just add a TODO here.

That said, don't we still use it for logging and getting leaseholder usage stats?  May be of note also that the `repl` argument is needed for any replicate queue planning function that considers any lease transfer, due to the allocator needing it in `ShouldTransferLease(..), TransferLeaseTargets(..), ValidLeaseTargets(..)`.  Which means I think `planAddOrReplace(Non)Voters(..)` are the only replicate queue functions it could be removed from.


<!-- Sent from Reviewable.io -->

kvoli reviewed (changes requested):
> Reviewed 1 of 1 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92172#-NHq_f0T0RGN5H4SrwTZ:-NJqHnTc6NW6M6RuQscD:b40owd2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7e54d176700707401226f43715f65ca3d60f48c/pkg/kv/kvserver/replicate_queue.go#L1220)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Could also just add a TODO here.

That said, don't we still use it for logging and getting leaseholder usage stats?  May be of note also that the `repl` argument is needed for any replicate queue planning function that considers any lease transfer, due to the allocator needing it in `ShouldTransferLease(..), TransferLeaseTargets(..), ValidLeaseTargets(..)`.  Which means I think `planAddOrReplace(Non)Voters(..)` are the only replicate queue functions it could be removed from.
</blockquote></details>

We can table it with a TODO for now.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 1587 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92172#-NJqIXiJ8FDa3BOi7_5y:-NJqIXiJ8FDa3BOi7_5z:b-h2gus4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5178580a4ad7f9632c42be3bbfca11854b8f7b24/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L1587)):*
> ```Go
> 		RaftStatus() *raft.Status
> 		GetFirstIndex() uint64
> 		Desc() *roachpb.RangeDescriptor
> ```

This is duplicated below now - we should remove it from the interface.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 1752 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92172#-NJqJlW_8VRuFSq2gH8D:-NJqJlW_8VRuFSq2gH8E:bpnbi5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/5178580a4ad7f9632c42be3bbfca11854b8f7b24/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L1752)):*
> ```Go
> 		RaftStatus() *raft.Status
> 		GetFirstIndex() uint64
> 		Desc() *roachpb.RangeDescriptor
> ```

ditto, this is duplicated below now - we should remove it from the interface.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 2006 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92172#-NJqJzRe1DcImYOB7HDw:-NJqJzRe1DcImYOB7HDx:bpnbi5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/5178580a4ad7f9632c42be3bbfca11854b8f7b24/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L2006)):*
> ```Go
> 		RaftStatus() *raft.Status
> 		GetFirstIndex() uint64
> 		Desc() *roachpb.RangeDescriptor
> ```

ditto, this is duplicated below now - we should remove it from the interface.


<!-- Sent from Reviewable.io -->
